File: LRU Cache - GFG/lru-cache.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// design the class in the most optimal way

struct node{
    int key;
    int val; 
    node* prev;
    node* next;
};
class LRUCache {
private: 
    unordered_map<int, node*> map;
    int cap;
    node* head;
    node* tail;
    
public:
    LRUCache(int capacity) {
        cap=capacity;
        head = new node();
        head->key=-1;
        head->val=-1; 
        
        tail= new node(); 
        tail->key=-1;
        tail->val=-1; 
    
        head->next=tail;
        tail->prev=head; 
    
    }
    
    int GET(int key) {
        if(map.find(key)==map.end())
            return -1;
        
        node* existingnode = map[key] ;
        int ret=existingnode->val;
        existingnode->prev->next  = existingnode->next ;
        existingnode->next->prev = existingnode->prev ;
        existingnode->prev=NULL;
        existingnode->next=NULL; 
        delete existingnode;
                
        node* headnext=head->next;
        node* newnode=new node();
        newnode->key=key;
        newnode->val=ret;
        head->next=newnode;
        newnode->prev=head;
        newnode->next = headnext; 
        headnext->prev=newnode;
        map[key] = newnode; 
        return ret;
    }
    
    void SET(int key, int val) { 
        if(map.find(key)!=map.end())
        {
            node *existingnode = map[key]; 
            existingnode->prev->next  = existingnode->next; 
            existingnode->next->prev = existingnode->prev ;
            existingnode->prev=NULL;
            existingnode->next=NULL;
            delete existingnode;
                
            node *headnext=head->next;
            node *newnode=new node();
            newnode->key=key;
            newnode->val=val;
            head->next=newnode ;
            newnode->prev=head ;
            newnode->next = headnext ;
            headnext->prev=newnode;
            map[key] = newnode; 
        } 
        
        else if (cap>map.size()){ 
            
            node *headnext=head->next; 
            node *newnode=new node();
            newnode->key=key;
            newnode->val=val;
            head->next=newnode; 
            newnode->prev=head;
            newnode->next = headnext; 
            headnext->prev=newnode;
            map[key] = newnode;
        }
            
        else{ 
            node *deletednode = tail->prev;
            deletednode->prev->next = tail;
            tail->prev=deletednode->prev;
            map.erase(deletednode->key); 
            deletednode->prev=NULL;
            deletednode->next=NULL;
            delete deletednode;
            
            node* headnext=head->next;
            node* newnode=new node();
            newnode->key=key;
            newnode->val=val;
            head->next=newnode; 
            newnode->prev=head ;
            newnode->next = headnext ;
            headnext->prev=newnode;
            map[key] = newnode;
        }
        
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

//{ Driver Code Starts.

int main()
{
    int t;
    cin >> t;
    while (t--)
    {

        int capacity;
        cin >> capacity;
        LRUCache *cache = new LRUCache(capacity);
        
        int queries;
        cin >> queries;
        while (queries--)
        {
            string q;
            cin >> q;
            if (q == "SET")
            {
                int key;
                cin >> key;
                int value;
                cin >> value;
                cache->SET(key, value);
            }
            else
            {
                int key;
                cin >> key;
                cout << cache->GET(key) << " ";
            }
        }
        cout << endl;
    }
    return 0;
}

// } Driver Code Ends

File: LRU Cache - GFG/lru-cache.py
class node :
    def __init__(self,key,val):
        self.key=key
        self.val=val
        self.prev=None
        self.next=None 
        
class LRUCache:
    def __init__(self, capacity):
        self.head=node(-1,-1) 
        self.tail=node(-1,-1) 
        self.head.next=self.tail 
        self.tail.prev=self.head 
        self.map={}
        self.cap=capacity 

    def get(self, key):
        if key not in self.map:
            return -1 
        existingnode = self.map[key] 
        ret=existingnode.val
        # print(ret)
        existingnode.prev.next  = existingnode.next 
        existingnode.next.prev = existingnode.prev 
        existingnode.prev=None 
        existingnode.next=None 
        del existingnode
                
        headnext=self.head.next 
        newnode=node(key,ret) 
        self.head.next=newnode 
        newnode.prev=self.head 
        newnode.next = headnext 
        headnext.prev=newnode
        self.map[key] = newnode 
        return ret

    def set(self, key,val):
        if key in self.map:
                existingnode = self.map[key] 
                existingnode.prev.next  = existingnode.next 
                existingnode.next.prev = existingnode.prev 
                existingnode.prev=None 
                existingnode.next=None 
                del existingnode
                
                headnext=self.head.next 
                newnode=node(key,val) 
                self.head.next=newnode 
                newnode.prev=self.head 
                newnode.next = headnext 
                headnext.prev=newnode
                self.map[key] = newnode 
                
        elif self.cap>len(self.map): 
            if key not in self.map :
                headnext=self.head.next 
                newnode=node(key,val) 
                self.head.next=newnode 
                newnode.prev=self.head 
                newnode.next = headnext 
                headnext.prev=newnode
                self.map[key] = newnode 
                # print(self.map[key], self.head.next.val)
            
        else: 
            deletednode = self.tail.prev 
            # print(self.head.next.val,deletednode.val)
            deletednode.prev.next = self.tail 
            self.tail.prev=deletednode.prev 
            del self.map[deletednode.key]
            deletednode.prev=None
            deletednode.next=None 
            del deletednode
            
            headnext=self.head.next 
            newnode=node(key,val) 
            self.head.next=newnode 
            newnode.prev=self.head 
            newnode.next = headnext 
            headnext.prev=newnode
            self.map[key] = newnode


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        cap = int(input())  # capacity of the cache
        qry=int(input())  #number of queries
        a = list(map(str, input().strip().split()))  # parent child info in list
        
        lru=LRUCache(cap)
        
       
        i=0
        q=1
        while q<=qry:
            qtyp=a[i]
            
            if qtyp=='SET':
                lru.set(int(a[i+1]),int(a[i+2]))
                i+=3
            elif qtyp=='GET':
                print(lru.get(int(a[i+1])),end=' ')
                i+=2
            q+=1
        print()
# } Driver Code Ends